drivers/mm/mm_drv_common.c

```diff
@@ -224,48 +287,62 @@ int sys_mm_drv_simple_remap_region(void *virt_old, size_t size,
 		uint8_t *va_new = (uint8_t *)virt_new + offset;
 		uintptr_t pa;
 		uint32_t flags;
-		int ret2;
 		bool to_map;
 
 		/*
-		 * va_old is mapped as checked above, so no need
-		 * to check for return value here.
+		 * Grab the physical address of the old mapped page
+		 * so the new page can map to the same physical address.
 		 */
-		(void)sys_mm_drv_page_phys_get(va_old, &pa);
+		ret = sys_mm_drv_page_phys_get(va_old, &pa);
+		if (ret != 0) {
+			__ASSERT(false, "cannot query %p\n", va_old);
 
-		to_map = true;
-		ret2 = sys_mm_drv_page_flag_get(va_old, &flags);
-		if (ret2 != 0) {
-			__ASSERT(false, "cannot query page %p\n", va_old);
+			/*
+			 * Reset the already mapped virtual addresses.
+			 * Note the offset is at the current failed address
+			 * which will not be included during unmapping.
+			 */
+			failure_mapping_reset(virt_new, offset);
 
-			ret = ret2;
-			to_map = false;
+			goto unlock_out;
 		}
 
-		ret2 = sys_mm_drv_unmap_page(va_old);
-		if (ret2 != 0) {
-			__ASSERT(false, "cannot unmap %p\n", va_old);
-
-			ret = ret2;
-		}
+		/*
+		 * Grab the flags of the old mapped page
+		 * so the new page can map to the same flags.
+		 */
+		to_map = true;
+		ret = sys_mm_drv_page_flag_get(va_old, &flags);
+		if (ret != 0) {
+			__ASSERT(false, "cannot query page %p\n", va_old);
 
-		if (!to_map) {
 			/*
-			 * Cannot retrieve flags of mapped virtual memory.
-			 * Skip mapping this page as we don't want to map
-			 * with unknown random flags.
+			 * Reset the already mapped virtual addresses.
+			 * Note the offset is at the current failed address
+			 * which will not be included during unmapping.
 			 */
-			continue;
+			failure_mapping_reset(virt_new, offset);
+
+			goto unlock_out;
 		}
 
-		ret2 = sys_mm_drv_map_page(va_new, pa, flags);
-		if (ret2 != 0) {
+		ret = sys_mm_drv_map_page(va_new, pa, flags);
+		if (ret != 0) {
 			__ASSERT(false, "cannot map 0x%lx to %p\n", pa, va_new);
 
-			ret = ret2;
+			/*
+			 * Reset the already mapped virtual addresses.
+			 * Note the offset is at the current failed address
+			 * which will not be included during unmapping.
+			 */
+			failure_mapping_reset(virt_new, offset);
+
+			goto unlock_out;
 		}
 	}
 
+	(void)unmap_locked(virt_old, size);
```

Comment:
Hardware should be able to map any number of virtual address to the same physical page.

If we cannot map the new region entirely, we need to roll back and remove any newly mapped addresses and also need to retain the old mappings. It is done this way because we do not have to re-map the old region. We "simply" remove the new mappings, and do not have to do rollback one by one. If the new mappings can be done entirely, we simply remove the whole old mapped region.